internal/pkg/client/cache/dir.go

```diff
@@ -22,69 +22,251 @@ const (
 	// for image downloads to be cached in
 	DirEnv = "SINGULARITY_CACHEDIR"
 
-	// RootDefault specifies the directory inside of ${HOME} that images are
+	// BasedirDefault specifies the directory inside of ${HOME} that images are
 	// cached in by default.
 	// Uses "~/.singularity/cache" which will not clash with any 2.x cache
 	// directory.
-	RootDefault = ".singularity/cache"
+	BasedirDefault = ".singularity"
+
+	// RootDefault is the default directory created in the base directory that will actually
+	// host the cache
+	RootDefault = "cache"
+
+	// StateInitialized represents the state of a give cache after successful initialization
+	StateInitialized = "initialized"
+
+	// StateInvalid represents the state of an invalid cache
+	StateInvalid = "invalid"
 )
 
-var root string
+// SingularityCache is an opaque structure representing a cache
+type SingularityCache struct {
+	// Basedir for the cache. This directory is never entirely deleted since it
+	// can be specified by the user via the DirEnv environment variable.
+	BaseDir string
+
+	// Root directory of the cache, within the basedir. This is the directory
+	// Singularity actually manages, i.e., that can safely be deleted as
+	// opposed to the base directory that is potentially managed (passed in)
+	// by the user
+	Root string
+
+	// State of the cache. We enable manual change of the state mainly for
+	// testing
+	State string
+
+	// Default specifies if the handle points at the default image cache or
+	// not. This enables quick lookup.
+	Default bool
+
+	// PreviousDirEnv stores the value of the DirEnv environment variable
+	// before it is implicitly modified. This is used to restore the
+	// environment configuration in some specific contexts.
+	PreviousDirEnv string
+
+	// Library provides the location of the Library cache
+	Library string
+
+	// OciTemp provides the location of the OciTemp cache
+	OciTemp string
+
+	// OciBlob provides the location of the OciBlob cache
+	OciBlob string
+
+	// Net provides the location of the Net cache
+	Net string
+
+	// Shub provides the location of the Shub cache
+	Shub string
+}
+
+// Create a new Singularity cache
+func Create() (*SingularityCache, error) {
+	// Singularity makes the following assumptions:
+	// - the default location for caches is specified by RootDefault
+	// - a user can specify the environment variable specified by DirEnv to change the location
+	// - a user can change the location of a cache at any time
+	// - but in the context of a Singularity command, the cache location cannot change once the command starts executing
+	basedir, err := getCacheBasedir()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get root of the cache: %s", err)
+	}
+
+	return Init(basedir)
+}
+
+// Init initializes a new cache within a given directory
+func Init(baseDir string) (*SingularityCache, error) {
```

Comment:
I'm trying to figure out if this should be exported or not. Gut feeling is "no". Is there a reason why it should?

Reply:
At the moment, I think you are right, it does not need to be exported. I will fix it.

Reply:
fixed